kimchi/src/circuits/polynomials/foreign_mul/gate.rs

```diff
@@ -0,0 +1,528 @@
+//! Foreign field multiplication gate
+
+use ark_ff::{FftField, Zero};
+use ark_poly::{univariate::DensePolynomial as DP, Evaluations as EV, Radix2EvaluationDomain as D};
+use array_init::array_init;
+use rand::{prelude::StdRng, SeedableRng};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::{Argument, ArgumentType},
+        constraints::ConstraintSystem,
+        domains::EvaluationDomains,
+        expr::{self, Column, E},
+        gate::{CircuitGate, GateType},
+        polynomial::COLUMNS,
+        wires::GateWires,
+    },
+    proof::ProofEvaluations,
+};
+
+use super::{ForeignMul0, ForeignMul1, ForeignMul2};
+
+impl<F: FftField> CircuitGate<F> {
+    /// Create foreign multiplication gate
+    pub fn create_foreign_mul(wires: &[GateWires; 8]) -> Vec<Self> {
+        vec![
+            /* Input: a */
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[0],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[1],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul1,
+                wires: wires[2],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul2,
+                wires: wires[3],
+                coeffs: vec![],
+            },
+            /* Input: b */
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[4],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[5],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul1,
+                wires: wires[6],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul2,
+                wires: wires[7],
+                coeffs: vec![],
+            },
+        ]
```

Comment:
two things:

* use `Wire::new(row)` to generate each wire
* add the necessary wiring here as well

Reply:
Makes sense, but for next iteration's PR